0015-authority.md

```diff
@@ -0,0 +1,478 @@
+        }
+
+        // The authority which could either be the owner or the server based on the authority mode uses the input to move.
+        if(HasAuthority)
+        {
+            if(m_MyInput.Value)
+            {
+                m_Rigidbody.AddForce(transform.up * m_Thrust);
+            }
+        }
+    }
+}
+```
+*The authority system allows NetworkVariables to be flexibly used as inputs.*
+
+## Changing Authority
+
+The `Authority` mode of a `NetworkObject` can be changed, on the server only, by calling `networkObject.ChangeAuthority(Authority.Server);`.
+
+This means which peer has authority also changes if the ownership of an object changes when in `owner` authority mode. This means the `HasAuthority` property should never be cached but always accessed directly.
+
+## OnGainedAuthority / OnLostAuthority
+
+`OnGainedAuthority` and `OnLostAuthority` functions can be overridden to get informed about changes to the `HasAuthority` status. `OnGainedAuthority` gets called when `HasAuthority` switches from false to true. `OnLostAuthority` gets called when it switches to false from true
+```csharp
+public override void OnGainedAuthority()
+{
+    Debug.Log("Authority Gained :)");
+}
+
+public override void OnLostAuthority()
+{
+    Debug.Log("Authority Lost :(");
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Why is Authority a mode?
+Why is authority not something like ownership which you give to a client. The reason for that is because authority is a relation between multiple peers. In an authoritative model a set of peers pass inputs to the authoritative peer for processing.
+
+Because we are a star topology we have some technical limitations. For instance we could not have one client generate the inputs but pass it to another client which holds authority (because this would break the encapsulation principle of time explained in the NetworkTime RFC).
+
+Because of the above we will only support authority models where the owner of an object collects inputs and sends them to the authoritative peer which has to be either the owner or the server.
+
+Accepting inputs from multiple peers is something we will not support with the current proposed design. The authority mode system allows for further extension to support those modes.
+
+## Ownership vs Authority
+This RFC will introduce some rules and responsibilities for each of the two systems.
+
+IsOwner:
+- Only one peer can own a NetworkObject.
+- The `owner` of a NetworkObject can use `ServerRpcs` on its `NetworkBehaviour` by default.
+- The `owner` is responsible for collecting inputs and redirecting them to the authority.
+- The `owner` is encouraged write to `NetworkVariables` with readpermission = authority but not to other NetworkVariables.
+- When the owner peer disconnects, the object it owns might get destroyed when `DontDestroyWithOwner` is set to false.
+
+HasAuthority:
```

Comment:
I'd be interested to put into writing the relation between IsOwner and HasAuthority here

Reply:
My guess is that it would end up in something like this:

Authority/Control gives a client possibility for input/output and temp. state (aka local write trough)
Owners have the final say about a state of an Object disregarding but maybe including Authority.

Scenarios for abstraction:
In context of movement this can give a client predictive abilities as it can set input/output and temp. state (client prediction) without being the owner(server-auth)
In context of an inventory this can make input-lag-less UI possible (anolog to client prediction) without being the owner(server-auth)

In those scenario cases either client or server could be the owner but it would not change the context.
Thus a correct abstraction away from isClient / isServer.

conclusion
A)  Maybe we need a more consistent naming?
-> Control vs Owner
-> Input Auth vs State Auth
-> Simulant vs Original (thus opening replica/copy nameing on the other maschines)

B) Networking Mode specific defaults?
-> Dedicated Servers would have Ownership over all objects but give Auth to specific clients.
-> Client/Host-Mode: Each client has both Ownership and Auth over specific Objects while Server is only messenger.

C) This would end up requiring either an Acknowledge or a state-rollback(like you commented already) if the owner decides against a change. Maybe we end up in something like Acknowledge First Authority VS Roll-Back Authority.
This RFC decided for a simple Authority check but misses out on the roll-back part (thus still requires rpcs)